php-frontend/src/main/java/org/sonar/php/parser/PHPLexicalGrammar.java

```diff
@@ -0,0 +1,454 @@
+    // fixme TOP_STATEMENT_LIST instead of EXPRESSION
+    b.rule(SCRIPT).is(FILE_OPENING_TAG, b.optional(EXPRESSION));
+
+    lexical(b);
+    punctuators(b);
+    keywords(b);
+
+    // FIXME COMPILATION_UNIT
+    b.setRootRule(EXPRESSION);
+
+    return b;
+  }
+
+  public static void lexical(LexerlessGrammarBuilder b) {
+    b.rule(SPACING).is(
+      b.skippedTrivia(b.regexp("[" + LexicalConstant.LINE_TERMINATOR + LexicalConstant.WHITESPACE + "]*+")),
+      b.zeroOrMore(
+        b.commentTrivia(b.regexp(LexicalConstant.COMMENT)),
+        b.skippedTrivia(b.regexp("[" + LexicalConstant.LINE_TERMINATOR + LexicalConstant.WHITESPACE + "]*+")))
+    ).skip();
+
+    // Literals
+    b.rule(HEREDOC).is(SPACING, b.regexp(LexicalConstant.HEREDOC));
+    b.rule(NUMERIC_LITERAL).is(SPACING, b.regexp(LexicalConstant.NUMERIC_LITERAL));
+//    b.rule(STRING_LITERAL).is(SPACING, b.firstOf(b.regexp(LexicalConstant.STRING_LITERAL), ENCAPS_STRING_LITERAL));
+
+    b.rule(STRING_WITH_ENCAPS_VAR_CHARACTERS).is(b.regexp(LexicalConstant.STRING_WITH_ENCAPS_VAR_CHARACTERS));
+//    b.rule(ENCAPS_STRING_LITERAL).is(SPACING, "\"", ENCAPS_LIST, "\"");
+
+    // Identifier
+    b.rule(ENCAPS_VAR_IDENTIFIER).is(
+      b.regexp("[" + LexicalConstant.WHITESPACE + "]*+"),
+      VAR_IDENTIFIER).skip();
+    b.rule(REGULAR_VAR_IDENTIFIER).is(SPACING, VAR_IDENTIFIER).skip();
+    b.rule(VAR_IDENTIFIER).is(b.regexp(LexicalConstant.VAR_IDENTIFIER));
+    b.rule(IDENTIFIER).is(SPACING, b.nextNot(KEYWORDS), b.regexp(LexicalConstant.IDENTIFIER));
+
+    // Tags & Inline HTML
+    b.rule(FILE_OPENING_TAG).is(SPACING, b.token(PHPTokenType.FILE_OPENING_TAG, b.regexp(LexicalConstant.PHP_START_TAG))).skip();
+    b.rule(INLINE_HTML).is(SPACING, b.token(PHPTokenType.INLINE_HTML, b.regexp(LexicalConstant.PHP_END_TAG))).skip();
+
+    b.rule(EOF).is(b.token(GenericTokenType.EOF, b.endOfInput())).skip();
+  }
+
+  private static void keywords(LexerlessGrammarBuilder b) {
+    Object[] rest = new Object[PHPKeyword.values().length - 2];
+
+    for (int i = 0; i < PHPKeyword.values().length; i++) {
+      PHPKeyword tokenType = PHPKeyword.values()[i];
+
+      // PHP keywords are case insensitive
+      b.rule(tokenType).is(SPACING, b.token(tokenType, b.regexp("(?i)" + tokenType.getValue())), b.nextNot(b.regexp(LexicalConstant.IDENTIFIER_PART))).skip();
+      if (i > 1) {
+        rest[i - 2] = b.regexp("(?i)" + tokenType.getValue());
+      }
+    }
+
+    b.rule(KEYWORDS).is(SPACING,
+      b.firstOf(
+        PHPKeyword.getKeywordValues()[0],
+        PHPKeyword.getKeywordValues()[1],
+        rest),
+      b.nextNot(b.regexp(LexicalConstant.IDENTIFIER_PART))
+    );
+  }
+
+  private static void punctuators(LexerlessGrammarBuilder b) {
+    for (PHPPunctuator p : PHPPunctuator.values()) {
+      b.rule(p).is(SPACING, b.token(p, p.getValue())).skip();
+    }
+  }
+
+  private static Object word(LexerlessGrammarBuilder b, String word) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Private method 'word' is never used. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedPrivateMethod)